### PR TITLE
Fix Ping.Unix.cs to handle various size of IPv4 header correctly

### LINKED ISSUE #21459: System.Net.Ping assumes fixed size IPv4 header for Unix
When [SendIcmpEchoRequestOverRawSocket()](https://github.com/dotnet/corefx/blob/master/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L43)  is used for ping in Linux, it assumes size of IPv4 header is fixed 20 bytes at [Ping.Unix.cs:19](https://github.com/dotnet/corefx/blob/master/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L19). 
In reality, size of IPv4 header can vary from 20 bytes to 60 bytes.

Therefore `SendIcmpEchoRequestOverRawSocket()` works incorrectly when IPv4 header is larger than 20 bytes and fails to find a correct echoed packet  due to wrong interpretation of a received packet.

This failure is observed when running `System.Net.Ping.Functional.Tests.dll` on environment where IPv4 header is larger than 20 bytes and 12 tests are failed as below.
```
      <test name="System.Net.NetworkInformation.Tests.PingTest.Sends_ReuseInstance_Hostname" type="System.Net.NetworkInformation.Tests.PingTest" method="Sends_ReuseInstance_Hostname" time="6.288208" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithHostAndTimeoutAndBuffer_Unix" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithHostAndTimeoutAndBuffer_Unix" time="10.0747374" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPings_ReuseInstance_Hostname" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPings_ReuseInstance_Hostname" time="5.0161132" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithHostAndTimeout" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithHostAndTimeout" time="10.0464476" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendAsyncs_ReuseInstance_Hostname" type="System.Net.NetworkInformation.Tests.PingTest" method="SendAsyncs_ReuseInstance_Hostname" time="5.0883178" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithHost" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithHost" time="5.0319823" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddressAndTimeoutAndBuffer_Unix" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithIPAddressAndTimeoutAndBuffer_Unix" time="10.0632629" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddress" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithIPAddress" time="5.0283813" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddressAndTimeout" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithIPAddressAndTimeout" time="10.0208434" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddressAndTimeoutAndBufferAndPingOptions_Unix" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithIPAddressAndTimeoutAndBufferAndPingOptions_Unix" time="10.0248107" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithIPAddress_AddressAsString" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithIPAddress_AddressAsString" time="5.0307922" result="Fail">
      <test name="System.Net.NetworkInformation.Tests.PingTest.SendPingAsyncWithHostAndTimeoutAndBufferAndPingOptions_Unix" type="System.Net.NetworkInformation.Tests.PingTest" method="SendPingAsyncWithHostAndTimeoutAndBufferAndPingOptions_Unix" time="10.0245666" result="Fail">
```

I'm not expert in network implementation and looking forward any advice and comments on this issue.

### PR DESCRIPTION
**Pull request by hqueue**
Fix #21459 

Size of IP header for IPv4 varies from 20 bytes to 60 bytes.

### CODE REVIEW

File: src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 1]
```diff
@@ -105,6 +106,17 @@ public partial class Ping
                     {
                         fixed (byte* bytesPtr = &receiveBuffer[0])
                         {
+                            if (isIpv4)
+                            {
+                                // Determine actual size of IP header
+                                byte ihl = (byte)(bytesPtr[0] & 0x0f); // Internet Header Length
```
Comment (mellinoe) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: very minor nit: Can you use `0x0F` (capital) to match other constants in this file?

**Comment by mellinoe** [comment]
@stephentoub Do you want to take a look as well?

File: src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs [hunk 2]
```diff
@@ -105,6 +106,17 @@ public partial class Ping
...
+                                ipHeaderLength = 4 * ihl;
+                                if (bytesReceived - ipHeaderLength < IcmpHeaderLengthInBytes)
+                                {
+                                    continue; // Not enough bytes to reconstruct actual IP header + ICMP header.
```
Comment (stephentoub) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: @mellinoe, I'm confused about the continue logic here.  It looks like we go back up and do another read, but we're reading back into receiveBuffer at position 0, and we don't seem to be tracking the total sum of everything that's been read, only what happened on the last read.  Am I reading that correctly?

Comment (mellinoe) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: The code currently assumes that an entire ICMP packet will be read into the buffer in the `ReceiveFromAsync` call, and if the buffer doesn't end up having what we need, then we retry. Thinking about it more, that is probably not valid, although I'm not sure of what the best practices are to handle it.

Comment (stephentoub) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: Ok. Well, this change doesn't seem to make that assumption worse, it just relies on it as well. We can go ahead with this, but we should probably revisit this code all-up to determine the right approach.

**Comment by stephentoub** [comment]
Thanks for the fix.

File: src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs (hunk 1 above)
Comment (hqueue) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: It seems that I'm late due to time zone difference.
Thank you for the review :)

File: src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs (hunk 2 above)
Comment (hqueue) [inline src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs]: I also had similar concerns when looking into this code first time, however I'm no expert in network and async I/O, so I just relied on the previous logic.

Thank you for the review and hope someone can improve this later :)

**Pull request merged**